Feature #164544632 – Add Solr support for private experiments

alfonsomunozpomer opened the pull request:
This is a multi-PR, multi-repo change.

# Requirements
Integration tests related to this story will fail if the following changes aren’t applied to the test data. All these changes have been applied in `lime` and the test data files used by Bamboo to have tests pass.

## Experiment files
Check out the [branch of the same name in the Expression Atlas Test Data repository](https://gitlab.ebi.ac.uk/amunoz/atlas-test-data/commits/feature/164544632-add-solr-support-for-private-experiments) (be aware that the branch checks out the `gxa` module to a branch of the same name, some `git submodules` magic might be needed). This branch adds the files of [E-GEOD-46858](https://www.ebi.ac.uk/gxa/experiments/E-GEOD-46858).

## PostgreSQL data
Private-import E-GEOD-46858 or download [`pgdata-gxa.tgz`](http://ftp.ebi.ac.uk/pub/databases/arrayexpress/data/atlas/test/pgdata-gxa.tgz), which contains E-GEOD-46858 as private.

## Solr
Create the new `bulk-analtyics-v1` collection. The scripts for this collection are in the [`solr-bulk` repository](https://github.com/ebi-gene-expression-group/solr-bulk). Run the analytics index endpoint and import the two private experiments. Atlernatively, download [solrcloud.tgz](http://ftp.ebi.ac.uk/pub/databases/arrayexpress/data/atlas/test/solrcloud.tgz).

# Changes
- Add `isPrivate` to the base experiment object. This field is populated from the field in the DTO that is created when reading the DB.
- Add a method to `ExperimentTrader` to retrieve an experiment without the `accesssKey`, disregarding whether it’s private or not.
- The analytics indexer uses the method above to read the bioentity identifiers and to stream the data points, which in turn adds the `is_private` field to the Solr documents. This means that `/admin/experiments/{experiment-accession}` can now index private experiments too. This safe since any operations under `/admin` require password authentication. The full reindex endpoint does so for public experiments only.
- Add the filter `is_private:false` to global searches but not to the experiment page search, to have full functionality in private experiments.

## Related PRs
- https://github.com/ebi-gene-expression-group/atlas-web-bulk/pull/2
- https://github.com/ebi-gene-expression-group/atlas-web-single-cell/pull/2

lingyun1010 commented on src/main/java/uk/ac/ebi/atlas/trader/ExperimentTrader.java:
```diff
@@ -22,8 +27,17 @@ public ExperimentTrader(ExperimentDao experimentDao,
+    }
+
+    @Cacheable(value="experimentByAccession", key="#experimentAccession")
+    public Experiment getExperiment(String experimentAccession) {
```
We should not cache the search results using the key as `experimentAcession`. It will be created when we import new private experiment as an admin and it can be accessed by the other users.

monicajianu commented on src/main/java/uk/ac/ebi/atlas/trader/ExperimentTrader.java:
```diff
@@ -22,10 +27,18 @@ public ExperimentTrader(ExperimentDao experimentDao,
 
-    @Cacheable(value="experimentByAccession", key="#experimentAccession")
+    @Cacheable(value="experimentByAccession")
+    public Experiment getExperiment(String experimentAccession, String accessKey) {
```
I think all these different getters make the class a little confusing. I can see a situation where someone could be tempted to just do `getExperiment` if they don't have an access key, thus accidentally retrieving private experiments. I think better naming could be used to protect against this.

alfonsomunozpomer commented on src/main/java/uk/ac/ebi/atlas/trader/factory/MicroarrayExperimentFactory.java:
```diff
@@ -56,6 +56,7 @@ public MicroarrayExperiment create(@NotNull ExperimentDto experimentDto,
                         .map(arrayDesignDao::getArrayDesign)
-                        .collect(Collectors.toList()));
+                        .collect(Collectors.toList()),
+                experimentDto.isPrivate());
```
Couldn’t see nothing wrong here...

alfonsomunozpomer commented on src/main/java/uk/ac/ebi/atlas/trader/factory/BaselineExperimentFactory.java:
```diff
@@ -61,7 +61,8 @@ public BaselineExperiment create(@NotNull ExperimentDto experimentDto,
                         factorsConfig.getMenuFilterFactorTypes(),
-                        factorsConfig.isOrderCurated()));
+                        factorsConfig.isOrderCurated()),
+                experimentDto.isPrivate());
```
Couldn’t see nothing wrong here...

alfonsomunozpomer commented:
@lingyun1010 I fixed the minor edits you suggested except a couple of places where indentation looked fine (?).

@monicajianu I had a chat with Lingyun and we renamed the `getExperiment(String experimentAccession` method. I’m not very satisfied with the result but it’s the best thing we could think of. I was thinking of something along the lines of `getExperimentOverridingPrivateFlag`, but Lingyun suggested adding the purpose of the method. Since the goal is to warn developers about not using that method on a general basis, the purpose very clearly establishes when to (and not to) use the method.
